src/components/validation/validators/component_spec.rs

```diff
@@ -68,15 +68,135 @@ impl Validator for ComponentSpecValidator {
             }
         }
 
-        // TODO: Check for the relevant telemetry events for the given component type.
-
-        Ok(vec![
+        let mut run_out = vec![
             format!(
                 "sent {} inputs and received {} outputs",
                 inputs.len(),
                 outputs.len()
             ),
             format!("received {} telemetry events", telemetry_events.len()),
-        ])
+        ];
+
+        let mut out = validate_telemetry(component_type, inputs, outputs, telemetry_events)?;
+        run_out.append(&mut out);
+
+        Ok(run_out)
+    }
+}
+
+fn validate_telemetry(
+    component_type: ComponentType,
+    inputs: &[TestEvent],
+    _outputs: &[Event],
+    telemetry_events: &[Event],
+) -> Result<Vec<String>, Vec<String>> {
+    let mut out: Vec<String> = Vec::new();
+    let mut errs: Vec<String> = Vec::new();
+
+    match component_type {
+        ComponentType::Source => {
+            match &mut validate_component_events_received(inputs, telemetry_events) {
+                Err(e) => errs.append(e),
+                Ok(m) => out.append(m),
+            }
```

Comment:
Just a note: simpler to use `Vec::extend` to append the success/failure results (including up on lines 80-81) since the only reason to use `Vec::append` would be if we wanted to keep the existing `Vec` and its underlying allocation.

Otherwise we just end up with noise here like the tacked on `&mut` in the match expression.

Reply:
Good call. Thanks!